include/tscore/HostLookup.h

```diff
@@ -23,109 +23,134 @@
 
 /*****************************************************************************
  *
- *  HostLookup.h - Interface to genernal purpose matcher
+ *  HostLookup.h - Interface to general purpose matcher
  *
  *
  ****************************************************************************/
 
 #pragma once
 
-#include "DynArray.h"
+#include <vector>
+#include <string>
+#include <string_view>
+#include <functional>
+#include <unordered_map>
 
 // HostLookup  constantss
-const int HOST_TABLE_DEPTH = 3; // Controls the max number of levels in the logical tree
-const int HOST_ARRAY_MAX   = 8; // Sets the fixed array size
+constexpr int HOST_TABLE_DEPTH = 3; // Controls the max number of levels in the logical tree
+constexpr int HOST_ARRAY_MAX   = 8; // Sets the fixed array size
 
-//
-//  Begin Host Lookup Helper types
-//
-enum HostNodeType {
-  HOST_TERMINAL,
-  HOST_HASH,
-  HOST_INDEX,
-  HOST_ARRAY,
-};
-enum LeafType {
-  LEAF_INVALID,
-  HOST_PARTIAL,
-  HOST_COMPLETE,
-  DOMAIN_COMPLETE,
-  DOMAIN_PARTIAL,
-};
+class CharIndex;
+class HostArray;
 
-// The data in the HostMatcher tree is pointers to HostBranches.
-//   No duplicates keys permitted in the tree.  To handle multiple
-//   data items bound the same key, the HostBranch has the lead_indexs
-//   array which stores pointers (in the form of array indexes) to
-//   HostLeaf structs
+// The data in the HostMatcher tree is pointers to HostBranches. No duplicates keys permitted in the tree.  To handle
+// multiple data items bound the same key, the HostBranch has the lead_indexs array which stores pointers (in the form
+// of array indexes) to HostLeaf structs
 //
-// There is HostLeaf struct for each data item put into the
-//   table
+// There is HostLeaf struct for each data item put into the table
 //
 struct HostLeaf {
-  LeafType type;
-  char *match;       // Contains a copy of the match data
-  int len;           // length of the data
-  bool isNot;        // used by any fasssst path ...
-  void *opaque_data; // Data associated with this leaf
+  /// Type of leaf.
+  enum Type {
+    LEAF_INVALID,
+    HOST_PARTIAL,
+    HOST_COMPLETE,
+    DOMAIN_COMPLETE,
+    DOMAIN_PARTIAL,
+  };
+  Type type{LEAF_INVALID};    ///< Type of this leaf instance.
+  std::string match;          // Contains a copy of the match data
+  bool isNot{false};          // used by any fasssst path ...
+  void *opaque_data{nullptr}; // Data associated with this leaf
+
+  HostLeaf() {}
+  HostLeaf(std::string_view name, void *data) : opaque_data(data)
+  {
+    if (!name.empty() && name.front() == '!') {
+      name.remove_prefix(1);
+      isNot = true;
+    } else {
+      isNot = false;
+    }
```

Comment:
isNot is set to false by default, so why is the else clause needed?

Reply:
Yeah, you are right.